### ISSUE TITLE
[css-content] Implementations and spec disagree regarding content: url on elements

### ISSUE THREAD

**Issue by emilio** [w3c/csswg-drafts#2889]
Per spec, we're supposed to wrap the content in a box, and thus this:

```html
<!doctype html>
<style>
div {
  content: url(..);
  border: 2px solid blue;
}
</style>
<div></div>
```

Is supposed to render a border on a block box, and the image as a replaced inline inside. Instead implementations replace the whole `<div>` with a replaced box.

**Comment by frivoal** [comment]
I don't think that spec is terribly reliable. It was completely abandoned for a while, and has been cleaned up somewhat, but I'd take the whole thing with a big grain of salt.

The editors (@fantasai @dauwhe ) would know more, but I suspect that for this spec, we're much more interested in aligning the spec with what implementations do (when we can figure out what that is) than in trying to push the implementations to match the spec.

**Comment by dauwhe** [comment]
> Is supposed to render a border on a block box, and the image as a replaced inline inside.

[The current draft says](https://drafts.csswg.org/css-content-3/#typedef-content-content-replacement):

>[using image in content property] makes the element or pseudo-element a replaced element, filled with the specified `image`. Its normal contents are suppressed and do not generate boxes, as if they were display: none.

But I'm probably missing something. As Florian said, this spec needs a lot of work. We cleaned out a lot of decades-old stuff a few years ago, but haven't yet had a chance to go through and look at everything from today's perspective.

**Comment by bzbarsky** [comment]
Note that UAs have different behavior for pseudo-elements and normal elements here, and the spec doesn't seem to reflect that.

**Comment by SelenIT** [comment]
Per the current draft, it seems that this different behavior for pseudo-elements seems to be a bug:

> Note: If the value of `<content-list>` is a single `<image>`, it must instead be interpreted as a `<content-replacement>`.

I assume that now browsers treat pseudo-elements with only one `<image>` and with several `<image>`s in the `content` value the same way, and don't apply any special formatting rules in the former case (and never did this before).

**Comment by emilio** [comment]
Yes, indeed I broke it in FF Nightly accidentally while implementing `content` for elements, see https://bugzilla.mozilla.org/show_bug.cgi?id=1473813.

**Comment by frivoal** [comment]
@dauwhe @fantasai (cc @astearns @atanassov ), given the known-to-be-unreliable state of the spec, how about adding one of these giant warnings to it, to make sure that any implementer looking at it treads with caution?

**Comment by SelenIT** [comment]
But according to the current draft, this “broken” behavior seems to be correct. The problem is that it might be not compatible with existing content, but it’s a different issue, isn’t it?

**Comment by fantasai** [comment]
The current status is that all implementations create a replaced element when `content: <image>` is applied to a normal element, and create an inline replaced child when `content: <image>` is applied to `::before`. [testcase](https://www.software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cstyle%3E%0Ap.one%2C%20p.two%3A%3Abefore%20%7B%20content%3A%20url(http%3A%2F%2Fxanthir.com%2Fpony)%3B%20border%3A%20solid%3B%20padding%3A%201em%3B%20display%3A%20block%3B%20%7D%0A%3C%2Fstyle%3E%0A%0A%3Cp%20class%3D%22one%22%3E%0A%3Cp%20class%3D%22two%22%3E)

Agenda+ to discuss:

- Whether we want to make an exception for `::before` and `::after` specifically, or for all pseudo-elements.
- Whether anyone wants to investigate the Web compatibility of migrating `::before` and `::after` to match regular elements.
